src/helics/core/mpi/MpiComms.cpp

```diff
@@ -9,38 +9,319 @@ Lawrence Livermore National Laboratory, operated by Lawrence Livermore National
 
 */
 #include "MpiComms.h"
-#include "helics/core/ActionMessage.h"
+#include "../../common/AsioServiceManager.h"
+#include "../ActionMessage.hpp"
 #include <memory>
 
+#include <mpi.h>
+
 namespace helics
 {
-MpiComms::MpiComms (const std::string &brokerTarget, const std::string &localTarget)
-    : CommsInterface (brokerTarget, localTarget)
+
+bool MpiComms::mpiCommsExists = false;
+std::mutex MpiComms::mpiSerialMutex;
+int MpiComms::commRank = -1;
+
+MpiComms::MpiComms ()
 {
+    if (mpiCommsExists)
+    {
+        std::cout << "WARNING: MPIComms object already created, unexpected results may occur" << std::endl;
+    }
+    std::cout << "MpiComms()" << std::endl;
+    initMPI();
+    mpiCommsExists = true;
+}
+
+MpiComms::MpiComms(const int &brokerRank)
+    : brokerRank(brokerRank)
+{
+    if (mpiCommsExists)
+    {
+        std::cout << "WARNING: MPIComms object already created, unexpected results may occur" << std::endl;
+    }
+    std::cout << "MpiComms(" << std::to_string(brokerRank) << ")" << std::endl;
+    initMPI();
+    if (brokerRank == -1)
+    {
+        setBrokerRank (commRank);
+    }
+    mpiCommsExists = true;
 }
+
 /** destructor*/
-MpiComms::~MpiComms () = default;
+MpiComms::~MpiComms ()
+{
+    std::cout << "Disconnecting MPIComms" << std::endl;
+    disconnect ();
+    std::cout << "Finished disconnect" << std::endl;
 
-void MpiComms::queue_rx_function () {}
+    int mpi_initialized;
+    MPI_Initialized(&mpi_initialized);
```

Comment:
does this call not have to be serialized?

Reply:
I'm actually not sure. MPI_Initialized is one of the few calls that can be used before MPI_Init gets called.

Reply:
I checked the MPI "standard", MPI_Initialized, MPI_Finalized, MPI_Query_Thread, MPI_Is_Thread_Main, MPI_Get_Version and MPI_Get_Library_Version are the only ones it says must always be thread-safe.